### PR TITLE
[DispatchCreation][LinalgExt] Add OnlineAttentionOp support in dispatch formation and reshape fusion

### PR DESCRIPTION
**Pull request by keshavvinayak01**
This PR teaches dispatch creation and the LinalgExt reshape-fusion infrastructure about OnlineAttentionOp, which is missing from the paths currently specialized only for `AttentionOp`. Torch input conversion was split out to #24177, and rewrite porting  was split to #24123. Codegen support landed in #24110.

### CODE REVIEW

**Review by keshavvinayak01 — COMMENTED** [review]
Refactored to build on top of https://github.com/iree-org/iree/pull/24110

File: compiler/plugins/input/Torch/InputConversion/test/attention.mlir [hunk 1]
```diff
@@ -10,119 +10,77 @@ func.func @attention(%arg0: tensor<5x2x3x4xf32>, %arg1: tensor<5x2x3x4xf32>, %ar
 // CHECK-DAG: #[[$MAP_V:.+]] = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d5, d3)>
```
Comment (IanWood1) [inline compiler/plugins/input/Torch/InputConversion/test/attention.mlir:10]: Note on testing: the large model tests in iree-test-suites don't contain torch attention ops (they use `iree_linalg_ext.attention` directly) so the test coverage of `iree_linalg_ext.online_attention` is pretty limited.


One potential way to test these changes would be to run CI and benchmarks after adding `ConvertAttentionToOnlineAttentionPass` early in the compiler. This would also catch missed fusion related regressions.

Comment (IanWood1) [inline compiler/plugins/input/Torch/InputConversion/test/attention.mlir:10]: Also testing related: should we duplicate the e2e tests for online attention?

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 2]
```diff
@@ -2056,7 +1954,7 @@ void OnlineAttentionOp::build(OpBuilder &odsBuilder, OperationState &odsState,
                               Value sum, ArrayAttr indexingMaps,
                               std::optional<Value> mask) {
   Value maskIn = mask.value_or(Value());
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1956]: Online attention needs to implement `LinalgFusionInterface`

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 3]
```diff
@@ -378,6 +378,29 @@ getReshapeInfo(LinalgExt::AttentionOp attentionOp) {
       });
 }
 
+static SmallVector<ReshapeOperandInfo>
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:381]: Can you add online attention tests for the transformations that now operate on online attention in LinalgExt & DispatchCreation?


Maybe it would be a good idea to split this up into several PRs that add support for online attention in these passes and then switch the torch lowering?

File: compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp [hunk 4]
```diff
@@ -211,8 +211,8 @@ static SmallVector<ReassociationIndices> getCollapsibleLoops(Operation *op) {
 
 /// Returns true if the given op is collapsible.
 static bool isEligibleForCollapse(Operation *op) {
```
Comment (IanWood1) [inline compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp:213]: This pass will crash because OnlineAttention does not implement the LinalgFusionInterface (evidence that tests are needed).

File: compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp [hunk 5]
```diff
@@ -211,8 +211,8 @@ static SmallVector<ReassociationIndices> getCollapsibleLoops(Operation *op) {
```
Comment (IanWood1) [inline compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp:211]: Some other places that might need support added: 
- `compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TransposeFusion.cpp`
- `compiler/src/iree/compiler/Preprocessing/Common/FoldAttentionWithTranspose.cpp` (not sure if this is still used)
- `compiler/src/iree/compiler/Codegen/Common/BlockDynamicDimensions.cpp`

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 6]
```diff
@@ -946,8 +969,8 @@ getOperandReassociation(AffineMap indexingMap,
 }
 
 /// Get the new value to use for a given `OpOperand` in the collapsed operation.
-static Value getCollapsedOpOperand(Location loc, AttentionOp op,
-                                   OpOperand *opOperand,
+template <typename OpTy>
+static Value getCollapsedOpOperand(Location loc, OpTy op, OpOperand *opOperand,
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:973]: nit:
```suggestion
template <typename AttentionTy>
static Value getCollapsedOpOperand(Location loc, AttentionTy op, OpOperand *opOperand,
```

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 7]
```diff
@@ -973,7 +996,8 @@ static Value getCollapsedOpOperand(Location loc, AttentionOp op,
       .getResult();
 }
 
-static void collapseOperandsAndResults(AttentionOp op,
+template <typename OpTy>
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:999]: same

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 8]
```diff
@@ -378,6 +378,29 @@ getReshapeInfo(LinalgExt::AttentionOp attentionOp) {
       });
 }
 
+static SmallVector<ReshapeOperandInfo>
+getReshapeInfo(LinalgExt::OnlineAttentionOp onlineAttentionOp) {
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:382]: Use template?

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 9]
```diff
@@ -1127,6 +1151,89 @@ collapseOpIterationDims(AttentionOp op,
...
+      inputOperands[1], inputOperands[2], inputOperands[3], outputOperands[0],
+      outputOperands[1], outputOperands[2],
+      rewriter.getAffineMapArrayAttr(indexingMaps), maskOperand);
+  rewriter.inlineRegionBefore(origOp.getRegion(), collapsedOp.getRegion(),
+                              collapsedOp.getRegion().begin());
+  return collapsedOp;
+}
+
+FailureOr<CollapseResult>
+collapseOpIterationDims(OnlineAttentionOp op,
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:1186]: Could this be templated?

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 10]
```diff
@@ -1127,6 +1151,89 @@ collapseOpIterationDims(AttentionOp op,
...
+  Value maskOperand;
+  if (inputOperands.size() > 4) {
+    maskOperand = inputOperands[4];
+  }
+
+  auto collapsedOp = OnlineAttentionOp::create(
+      rewriter, origOp.getLoc(), resultTypes, inputOperands[0],
+      inputOperands[1], inputOperands[2], inputOperands[3], outputOperands[0],
+      outputOperands[1], outputOperands[2],
+      rewriter.getAffineMapArrayAttr(indexingMaps), maskOperand);
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:1179]: This `::create` and the usasage in `createCollapsedOp` for `LinalgExt::Attention` should be using this builder overload:

https://github.com/iree-org/iree/blob/f3220d8f29bcdc4ad7067052a14a0d51a81f5458/compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp#L1166-L1170

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp [hunk 11]
```diff
@@ -1127,6 +1151,89 @@ collapseOpIterationDims(AttentionOp op,
...
+        return getCollapsedOpIndexingMap(map, collapsingInfo);
+      }));
+
+  SmallVector<utils::IteratorType> iteratorTypes(getCollapsedOpIteratorTypes(
+      origOp.getLoopIteratorTypes(), collapsingInfo));
+
+  Value maskOperand;
+  if (inputOperands.size() > 4) {
+    maskOperand = inputOperands[4];
+  }
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:1173]: If you use the other builder (previous comment) you shouldn't need this special mask handling & you can probably template these.

**Review by IanWood1 — COMMENTED** [review]
One concern that I have is that I don't think we will be currently able to fuse `iree_linalg_ext.online_attention` -> `linalg.generic` (1/sum + x) into a single dispatch. This will end up regressing a lot of our benchmarks.

File: compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp (hunk 4 above)
Comment (keshavvinayak01) [reply compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp:213]: Yes, I’m waiting for Kunwar’s PR to merge before I merge this, I’ll rebase first.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp (hunk 10 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp:1179]: Opened https://github.com/iree-org/iree/pull/24128 for this because there was other code to cleanup, too.

**Review by hanhanW — DISMISSED** [review]
> Add end-to-end OnlineAttentionOp support

This is lacking tests. Where is the evidence of the claim? The only test I see is the frontend lowering (i.e., torch -> linalg_ext). I think you at least need new tests in [attention.mlir](https://github.com/iree-org/iree/blob/main/tests/e2e/linalg_ext_ops/attention.mlir), and please revisit the tests for rest of changes.

File: compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp (hunk 5 above)
Comment (keshavvinayak01) [reply compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp:211]: TransposeFusion.cpp is addressed in #24123 (the rewrite-templatization PR split out from this one). After that lands, both `FuseTransposeWithAttentionLikeOp` and `BubbleTransposeVFromAttentionLikeOp` run on `OnlineAttentionOp`

We can merge that PR first.

**Comment by keshavvinayak01** [comment]
Separated out the [TMTensor lowering](https://github.com/iree-org/iree/pull/24177) to another PR.

**Review by keshavvinayak01 — COMMENTED** [review]
Bifurcated the PR into smaller chunks

**Review by IanWood1 — APPROVED** [review]

**Comment by keshavvinayak01** [comment]
@hanhanW Could you re-review?
